### PR TITLE
Implement APIs for CampaignReport

### PR DESCRIPTION
**Pull request by cungbac**
1. Create CampaignReport
Path: https://localhost:7066/api/v1/campaign-rating/create
Method: POST
Body:

```
{
    "fromDateTime": "2024-10-12",
    "toDateTime": "2024-12-12",
    "title": "cash",
    "subtitle": "",
    "summary": "",
    "detail": "",
    "amount": 1000,
    "unitOfMeasurement": "vnd",
    "documentsUrl": "",
    "campaignId": "23c0f172-f38f-4077-bf0a-888868b79974"
}
```
Response

```
{
    "id": "22fb7a12-2324-481c-bedc-00162efc18c9",
    "fromDateTime": "2024-10-12T00:00:00",
    "toDateTime": "2024-12-12T00:00:00",
    "title": "cash",
    "subtitle": "",
    "summary": "",
    "detail": "",
    "amount": 1000,
    "unitOfMeasurement": "vnd",
    "documentsUrl": "",
    "campaignId": "23c0f172-f38f-4077-bf0a-888868b79974",
    "createdAt": "2024-11-13T15:26:24.8162218Z",
    "createdBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505",
    "updatedAt": "2024-11-13T15:26:24.8162218Z",
    "updatedBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505"
}
```

2. Get CampaignReport by Id
Path: https://localhost:7066/api/v1/campaign-report/{campaignReportId}
Method: GET
Body: None
Response
```
{
    "id": "22fb7a12-2324-481c-bedc-00162efc18c9",
    "fromDateTime": "2024-10-11T17:00:00Z",
    "toDateTime": "2024-12-11T17:00:00Z",
    "title": "cash",
    "subtitle": "",
    "summary": "",
    "detail": "",
    "amount": 1000,
    "unitOfMeasurement": "vnd",
    "documentsUrl": "",
    "campaignId": "23c0f172-f38f-4077-bf0a-888868b79974",
    "createdAt": "2024-11-13T15:26:24.816221Z",
    "createdBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505",
    "updatedAt": "2024-11-13T15:26:24.816221Z",
    "updatedBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505"
}
```

3. Update CampaignReport
Path: https://localhost:7066/api/v1/campaign-report/{campaignReportId}
Method: PUT
Body:
```
{
    "id": "22fb7a12-2324-481c-bedc-00162efc18c9",
    "fromDateTime": "2024-10-12",
    "toDateTime": "2024-12-12",
    "title": "cash",
    "subtitle": "",
    "summary": "",
    "detail": "",
    "amount": 1000,
    "unitOfMeasurement": "vnd",
    "documentsUrl": "",
    "campaignId": "23c0f172-f38f-4077-bf0a-888868b79974"
}
```
Response:

```
{
    "id": "22fb7a12-2324-481c-bedc-00162efc18c9",
    "fromDateTime": "2024-10-12T00:00:00",
    "toDateTime": "2024-12-12T00:00:00",
    "title": "cash",
    "subtitle": "",
    "summary": "",
    "detail": "",
    "amount": 1000,
    "unitOfMeasurement": "vnd",
    "documentsUrl": "",
    "campaignId": "23c0f172-f38f-4077-bf0a-888868b79974",
    "createdAt": "2024-11-13T15:26:24.816221Z",
    "createdBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505",
    "updatedAt": "2024-11-13T15:29:03.7052593Z",
    "updatedBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505"
}
```
4. Delete CampaignReport
Path: https://localhost:7066/api/v1/campaign-report/{campaignReportId}
Method: DELETE
Body:
```
{
    "reasonForDeletion": "No longer needed"
}
```
Response:
```
{
    "id": "22fb7a12-2324-481c-bedc-00162efc18c9",
    "fromDateTime": "2024-10-11T17:00:00Z",
    "toDateTime": "2024-12-11T17:00:00Z",
    "title": "cash",
    "subtitle": "",
    "summary": "",
    "detail": "",
    "amount": 1000,
    "unitOfMeasurement": "vnd",
    "documentsUrl": "",
    "campaignId": "23c0f172-f38f-4077-bf0a-888868b79974",
    "createdAt": "2024-11-13T15:26:24.816221Z",
    "createdBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505",
    "deletedAt": "2024-11-13T15:30:16.7104522Z",
    "deletedBy": "f5c2063e-9a14-4ab4-bce2-7f5117ad5505",
    "reasonForDeletion": "No longer needed"
}
```

### CODE REVIEW

File: src/DonateHope.Core/Services/CampaignReportServices/CampaignReportDeleteService.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                "Failed to delete campaign report {CampaignReportId}. Error: {ErrorMessage}",
+                campaignReportId,
+                deletedResult.Errors.First().Message
+                );
+            return new ProblemDetailsError("Failed to delete campaign report.");
+        }
+        
+        _logger.LogInformation(
+            "Successfully deleted campaign report {CampaignReportId}", campaignReportId);
+        return _campaignReportMapper.MapCampaignReportToCampaignReportDeleteResponseDto(deletedCampaignReport);
```
Comment (AnNguyenLe) [inline src/DonateHope.Core/Services/CampaignReportServices/CampaignReportDeleteService.cs]: for deletion, we just want to know if the delete action succeeded or not. We don't have to return any value. Result.Ok() should be enough

Comment (AnNguyenLe) [reply src/DonateHope.Core/Services/CampaignReportServices/CampaignReportDeleteService.cs]: also. please remove the the lines of code related to returning deleted value here:
Such as:
at line 45:
deletedCampaignReport.DeletedAt = DateTime.UtcNow;
deletedCampaignReport.DeletedBy = deletedBy;
deletedCampaignReport.ReasonForDeletion = reasonForDeletion;

At mapper for deletion: MapCampaignReportToCampaignReportDeleteResponseDto

File: src/DonateHope.Core/Mappers/CampaignReportMapper.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    [MapperIgnoreTarget(nameof(CampaignReport.CreatedBy))]
+    [MapperIgnoreTarget(nameof(CampaignReport.IsDeleted))]
+    [MapperIgnoreTarget(nameof(CampaignReport.DeletedAt))]
+    [MapperIgnoreTarget(nameof(CampaignReport.DeletedBy))]
+    [MapperIgnoreTarget(nameof(CampaignReport.CampaignId))]
+    public partial CampaignReport MapCampaignReportUpdateRequestDtoToCampaignReport(CampaignReportUpdateRequestDto campaignReportUpdateRequestDto);
+    
+    [MapperIgnoreSource(nameof(campaignReport.UpdatedAt))]
+    [MapperIgnoreSource(nameof(campaignReport.UpdatedBy))]
+    public partial CampaignReportDeleteResponseDto MapCampaignReportToCampaignReportDeleteResponseDto(CampaignReport campaignReport);
```
Comment (AnNguyenLe) [inline src/DonateHope.Core/Mappers/CampaignReportMapper.cs]: No need mapper for delete response

File: src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            updateRequestDto,
+            parsedUserId
+        );
+
+        if (updatedResult.IsFailed)
+        {
+            return updatedResult.Errors.ToDetailedBadRequest();
+        }
+        
+        return CreatedAtRoute(
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs]: For PUT request, returning 200 should be enough

File: src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            campaignReportId,
+            deletedBy,
+            reasonForDeletionRequestDto.ReasonForDeletion
+            );
+
+        if (result.IsFailed)
+        {
+            return result.Errors.ToDetailedBadRequest();
+        }
+        return result.Value;
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs]: For DELETE request, return 200 or 204 should be enough. No need return any value

File: src/DonateHope.WebAPI/StartupExtensions/DependencyInjectionServicesExtension.cs [hunk 5]
```diff
@@ -64,6 +67,14 @@ IConfiguration configuration
...
+        services.TryAddSingleton<CampaignReportMapper>();
+        services.TryAddScoped<ICampaignReportCreateService, CampaignReportCreateService>();
+        services.TryAddScoped<ICampaignReportRetrieveService, CampaignReportRetrieveService>();
+        services.TryAddScoped<ICampaignReportUpdateService, CampaignReportUpdateService>();
+        services.TryAddScoped<ICampaignReportDeleteService, CampaignReportDeleteService>();
+        services.AddValidatorsFromAssemblyContaining<CampaignReportDeleteRequestValidator>();
+        
         return services;
     }
 }
```
Comment (AnNguyenLe) [inline src/DonateHope.WebAPI/StartupExtensions/DependencyInjectionServicesExtension.cs:90]: Thank for your hard work!

**Review by AnNguyenLe — CHANGES_REQUESTED** [review]

File: src/DonateHope.WebAPI/StartupExtensions/DependencyInjectionServicesExtension.cs (hunk 5 above)
Comment (cungbac) [reply src/DonateHope.WebAPI/StartupExtensions/DependencyInjectionServicesExtension.cs:90]: Thanks for your review

File: src/DonateHope.Core/Services/CampaignReportServices/CampaignReportDeleteService.cs (hunk 1 above)
Comment (cungbac) [reply src/DonateHope.Core/Services/CampaignReportServices/CampaignReportDeleteService.cs]: removed those line of codes

File: src/DonateHope.Core/Mappers/CampaignReportMapper.cs (hunk 2 above)
Comment (cungbac) [reply src/DonateHope.Core/Mappers/CampaignReportMapper.cs]: mapper deleted

File: src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs (hunk 3 above)
Comment (cungbac) [reply src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs]: returned 200 only

File: src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs (hunk 4 above)
Comment (cungbac) [reply src/DonateHope.WebAPI/Controllers/v1/CampaignReport/CampaignReportController.cs]: returned 200 only

**Review by AnNguyenLe — APPROVED** [review]

**Pull request merged**